crates/store/src/db/models/queries/transactions.rs

```diff
@@ -126,14 +174,203 @@ pub struct TransactionSummaryRowInsert {
+    let mut last_transaction_id: Option<Vec<u8>> = None;
+
+    loop {
+        let mut query =
+            SelectDsl::select(schema::transactions::table, TransactionRecordRaw::as_select())
+                .filter(schema::transactions::block_num.gt(block_range.start().to_raw_sql()))
+                .filter(schema::transactions::block_num.le(block_range.end().to_raw_sql()))
+                .filter(schema::transactions::account_id.eq_any(&desired_account_ids))
+                .into_boxed();
+
+        // Apply cursor-based pagination using the last seen (block_num, transaction_id)
+        if let (Some(last_block), Some(last_tx_id)) = (last_block_num, &last_transaction_id) {
+            query = query.filter(
+                schema::transactions::block_num
+                    .gt(last_block)
+                    .or(schema::transactions::block_num
+                        .eq(last_block)
+                        .and(schema::transactions::transaction_id.gt(last_tx_id))),
+            );
+        }
+
+        let chunk = query
+            .order((
+                schema::transactions::block_num.asc(),
+                schema::transactions::transaction_id.asc(),
+            ))
+            .limit(NUM_TXS_PER_CHUNK)
+            .load::<TransactionRecordRaw>(conn)
+            .map_err(DatabaseError::from)?;
+
+        // If no more data, we're done
+        if chunk.is_empty() {
+            break;
+        }
+
+        // Add transactions from this chunk one by one until we hit the limit
+        let mut added_from_chunk = 0;
+        for tx in chunk {
+            if total_size + tx.size_in_bytes <= MAX_PAYLOAD_BYTES {
+                total_size += tx.size_in_bytes;
+                // Update cursor position for next iteration
+                last_block_num = Some(tx.block_num);
+                last_transaction_id = Some(tx.transaction_id.clone());
+                all_transactions.push(tx);
+                added_from_chunk += 1;
+            } else {
+                // Can't fit this transaction, stop here
+                break;
+            }
+        }
+
+        // Break if chunk incomplete (size limit hit or data exhausted)
+        if added_from_chunk < NUM_TXS_PER_CHUNK {
+            break;
         }
     }
+
+    // Ensure block consistency: remove the last block if it's incomplete
+    // (we may have stopped loading mid-block due to size constraints)
+    let mut filtered_raw = all_transactions;
+
+    if total_size >= MAX_PAYLOAD_BYTES {
+        if let Some(last_tx) = filtered_raw.last() {
+            let last_block_num = last_tx.block_num;
+            filtered_raw.retain(|tx| tx.block_num != last_block_num);
+        }
+    }
+
+    // Convert to transaction records
+    let headers: Vec<crate::db::TransactionRecord> = vec_raw_try_into(filtered_raw)?;
+
+    let last_included_block = headers.last().map_or(*block_range.end(), |tx| tx.block_num);
+
+    Ok((last_included_block, headers))
```

Comment:
Related to the previous comments - can we not simplify this block to look something like:
```rust
if total_size >= MAX_PAYLOAD_BYTES {
    let last_block_num = last_block_num.expect("todo");
    let filtered_transactions = vec_raw_try_into(
        all_transactions.into_iter().take_while(|row| row.block_num != last_block_num),
    )?;

    let last_included_block = BlockNumber::from_raw_sql(last_block_num.saturating_sub(1))?;
    Ok((last_included_block, filtered_transactions))
} else {
    let last_included_block = (last_block_num.unwrap_or(*block_range.end()) as u32).into();
    Ok((last_included_block, vec_raw_try_into(all_transactions)?))
}
```

If this works, it also has the benefit of iterating over the result set only once in both cases.